doc/dev/transforms.rst

```diff
@@ -0,0 +1,594 @@
+        %h = arith.constant 0.1 : f64
+
+        %fres = func.call @my_func(%x, %y, %z) : (f64, f64, f64) -> f64
+
+        %xph = arith.addf %x, %h : f64
+        %fxph = func.call @my_func(%xph, %y, %z) : (f64, f64, f64) -> f64
+        %diffx = arith.subf %fxph, %fres : f64
+        %dx = arith.divf %diffx, %h
+
+        %yph = arith.addf %y, %h : f64
+        %fyph = func.call @my_func(%x, %yph, %z) : (f64, f64, f64) -> f64
+        %diffy = arith.subf %fyph, %fres : f64
+        %dy = arith.divf %diffy, %h
+
+        %zph = arith.addf %z, %h : f64
+        %fzph = func.call @my_func(%x, %y, %zph) : (f64, f64, f64) -> f64
+        %diffz = arith.subf %fzph, %fres : f64
+        %dz = arith.divf %diffz, %h
+
+        func.return %dx, %dy, %dz : f64, f64, f64
+    }
+
+Finally, we have to amend our rewrite function to invoke the new function we created and delete the
+``GradOp`` from the IR:
+
+.. code-block:: cpp
+
+    void FiniteDiffLowering::rewrite(GradOp op, PatternRewriter &rewriter)
+    {
+        ...
+            populateFiniteDiffMethod(rewriter, op, gradFn);
+        }
+
+        rewriter.replaceOpWithNewOp<func::CallOp>(op, gradFn, op.getArgOperands());
+    }
+
+Note how we can create a new operation, take its results, and use those to replace another operation
+in one go. This turns the previous IR:
+
+.. code-block::
+
+    func.func @my_func(%x: f64, %y: f64, %z: f64) -> f64 {
+        ...
+    }
+
+    func.func @my_func.finitediff(f64, f64, f64) -> (f64, f64, f64) {
+        <contents of populateFiniteDiffMethod>
+    }
+
+    %deriv:3 = gradient.grad "fd" @my_func(%x, %y, %z) : (f64, f64, f64) -> (f64, f64, f64)
+
+into:
+
+.. code-block::
+
+    func.func @my_func(%x: f64, %y: f64, %z: f64) -> f64 {
+        ...
+    }
+
+    func.func @my_func.finitediff(f64, f64, f64) -> (f64, f64, f64) {
+        <contents of populateFiniteDiffMethod>
+    }
+
+    %deriv:3 = func.call @my_func.finitediff(%x, %y, %z) : (f64, f64, f64) -> (f64, f64, f64)
+
+Catalyst's Transformation Library
+=================================
+
+Why don't you try writing a pass of your own? Or have a look at our existing transformations from
+
+- the `quantum dialect <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/lib/Quantum/Transforms>`_,
+- the `gradient dialect <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/lib/Gradient/Transforms>`_,
+- or the `catalyst utility dialect <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/lib/Catalyst/Transforms>`_.
+
+The pass declarations and headers for transformations are located in the include directory of each
+dialect: `quantum <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/include/Quantum/Transforms>`_,
+`gradient <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/include/Gradient/Transforms>`_,
+and `catalyst <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/include/Catalyst/Transforms>`_.
```

Comment:
This is great 💯 You get to the end, and it makes you excited to try building something